src/main/java/jenkins/plugins/git/GitBranchSCMHead.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.plugins.git;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.Extension;
+import jenkins.scm.api.SCMHead;
+import jenkins.scm.api.SCMHeadMigration;
+import jenkins.scm.api.SCMRevision;
+import org.eclipse.jgit.lib.Constants;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Represents a Git Branch in {@code refs/origin/} (but uses
+ * {@link AbstractGitSCMSource#REF_SPEC_REMOTE_NAME_PLACEHOLDER_STR} so that it correctly handles when the remote
+ * name is customized.)
+ *
+ * @since 3.7.0
+ */
+public class GitBranchSCMHead extends SCMHead implements BranchSpecSCMHead {
+
+    /**
+     * Constructor.
+     *
+     * @param name the name.
+     */
+    public GitBranchSCMHead(@NonNull String name) {
+        super(name);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String getBranchSpec() {
+        return Constants.R_REFS + AbstractGitSCMSource.REF_SPEC_REMOTE_NAME_PLACEHOLDER_STR + "/" + getName();
```

Comment:
It may make more sense to have this be `Constants.R_HEADS + getName()`

(we still will want the behaviour of any `@{remote}` strings being replaced by the remote name on the return value of `getBranchSpec()`)

The question here is really what makes it easier for `GitSCMBuilder` to determine the minimal refspec to pull

Reply:
I am thinking that we may need a second method to determine the remote ref spec...

So by way of explanation...

In the `git ls-remote` the master branch will be reported as `refs/heads/master`, we will then produce a refspec template for fetching which is `+refs/heads/master:refs/@{remote}/master` and then in the normal case that will be turned into the refspec of `+refs/heads/master:refs/remotes/origin/master`

We then need the `GitSCM`'s `BranchSpec` to be `refs/remotes/origin/master` so that we actually checkout the correct master branch (as `master` could cause confusion if the user configured additional refspecs and they contained e.g. `refs/remotes/fork/master`... and worse still if the local workspace has already checked out a local `master` as then `refs/heads/master` will win in priority... we only get saved from serious issues because we then specify the revision to checkout and because we have fetched the branch with the revision it all *magically* seems to work... but a manually triggered  build could fall foul)

What I don't know is if the change in `BranchSpec` from `master` to `refs/remotes/origin/master` will have any other side impact...

oh and I thing have a :bug: as this is returning `refs/origin/master` and not `refs/remotes/origin/master`